baseapp/streaming.go

```diff
@@ -0,0 +1,111 @@
+package baseapp
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+
+	"cosmossdk.io/store/streaming"
+	storetypes "cosmossdk.io/store/types"
+	"github.com/spf13/cast"
+
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	servertypes "github.com/cosmos/cosmos-sdk/server/types"
+)
+
+const (
+	StreamingTomlKey                  = "streaming"
+	StreamingABCITomlKey              = "abci"
+	StreamingABCIPluginTomlKey        = "plugin"
+	StreamingABCIKeysTomlKey          = "keys"
+	StreamingABCIStopNodeOnErrTomlKey = "stop-node-on-err"
+)
+
+// RegisterStreamingServices registers streaming services with the BaseApp.
+func (app *BaseApp) RegisterStreamingServices(appOpts servertypes.AppOptions, keys map[string]*storetypes.KVStoreKey) {
+	// register streaming services
+	streamingCfg := cast.ToStringMap(appOpts.Get(StreamingTomlKey))
+	for service := range streamingCfg {
+		pluginKey := fmt.Sprintf("%s.%s.%s", StreamingTomlKey, service, StreamingABCIPluginTomlKey)
+		pluginName := strings.TrimSpace(cast.ToString(appOpts.Get(pluginKey)))
+		if len(pluginName) > 0 {
+			logLevel := cast.ToString(appOpts.Get(flags.FlagLogLevel))
+			plugin, err := streaming.NewStreamingPlugin(pluginName, logLevel)
+			if err != nil {
+				app.logger.Error("failed to load streaming plugin", "error", err)
+				os.Exit(1)
+			}
+			if err := app.registerStreamingPlugin(appOpts, keys, plugin); err != nil {
+				app.logger.Error("failed to register streaming plugin", "error", err)
+				os.Exit(1)
+			}
+		}
+	}
+}
+
+// registerStreamingPlugin registers streaming plugins with the BaseApp.
+func (app *BaseApp) registerStreamingPlugin(
+	appOpts servertypes.AppOptions,
+	keys map[string]*storetypes.KVStoreKey,
+	streamingPlugin interface{},
+) error {
+	switch t := streamingPlugin.(type) {
+	case storetypes.ABCIListener:
+		app.registerABCIListenerPlugin(appOpts, keys, t)
+	default:
+		return fmt.Errorf("unexpected plugin type %T", t)
+	}
+	return nil
```

Comment:
This would read easier w/o the switch, i.e.

```go
v, ok := streamingPlugin.(storetypes.ABCIListener)
if !ok {
  // ...
}

app.registerABCIListenerPlugin(appOpts, keys, v)
return nil
```

Reply:
changed to above.